core/shared/src/main/scala/fs2/async/mutable/Queue.scala

```diff
@@ -160,20 +180,56 @@ object Queue {
         def timedEnqueue1(a: A, timeout: FiniteDuration, scheduler: Scheduler): F[Boolean] = offer1(a)
         def offer1(a: A): F[Boolean] =
           qref.modify { s =>
-            if (s.deq.isEmpty) s.copy(queue = s.queue :+ a)
-            else s.copy(deq = s.deq.tail)
+            if (s.deq.isEmpty) s.copy(queue = s.queue :+ a, peek = None)
+            else s.copy(deq = s.deq.tail, peek = None)
           }.flatMap { c =>
-            if (c.previous.deq.isEmpty) // we enqueued a value to the queue
-              signalSize(c.previous, c.now).as(true)
-            else // queue was empty, we had waiting dequeuers
-              c.previous.deq.head.setAsyncPure(Chunk.singleton(a)).as(true)
+            val dq = if (c.previous.deq.isEmpty) {
+              // we enqueued a value to the queue
+              signalSize(c.previous, c.now)
+            } else {
+              // queue was empty, we had waiting dequeuers
+              c.previous.deq.head.setAsyncPure(Chunk.singleton(a))
+            }
+            val pk = if (c.previous.peek.isEmpty) {
+              // no peeker to notify
+              F.unit
+            } else {
+              // notify peekers
+              c.previous.peek.get.setAsyncPure(a)
+            }
+            (dq *> pk).as(true)
           }
 
         def dequeue1: F[A] = cancellableDequeue1.flatMap { _._1 }
 
+        def peek1: F[A] = peek1Impl.flatMap {
+          case Left(ref) => ref.get
+          case Right(a) => F.pure(a)
+        }
+
         def timedDequeue1(timeout: FiniteDuration, scheduler: Scheduler): F[Option[A]] =
           timedDequeueBatch1(1, timeout, scheduler).map(_.map(_.strict.head))
 
+        def timedPeek1(timeout: FiniteDuration, scheduler: Scheduler): F[Option[A]] = {
+          peek1Impl.flatMap {
+            case Left(ref) => ref.timedGet(timeout, scheduler)
+            case Right(a) => F.pure(Some(a))
+          }
+        }
+
+        private def peek1Impl: F[Either[Ref[F, A], A]] = ref[F, A].flatMap { ref =>
+          qref.modify2 { state =>
+            if (state.queue.isEmpty) {
+              state.peek match {
+                case None => (state.copy(peek = Some(ref)), Left(ref))
+                case Some(r) => (state, Left(r))
+              }
+            } else {
+              (state, Right(state.queue.head))
+            }
+          }.map(_._2)
+        }
+
```

Comment:
Generally speaking I like the approach here, not sure whether we should go the full route of having a `cancellablePeek1` like for `dequeue` or not

Reply:
I'll think about it.